test/guiding_test.jl

```diff
@@ -556,14 +543,118 @@ end
   Z = pCN(solfw1.W, 1.0)
-  solfw3, ll3 = MSDE.forwardguiding(kg3, message, (x0, ll0), Z)
+  solfw3, ll3 = MSDE.forwardguiding(kg3, message, (x0, ll0), EM(false), Z)
   @test ll1 ≈ ll3
   @test isapprox(solfw1.u, solfw3.u, rtol=1e-14)
 
   # inplace=false
   Z = pCN(solfw1.W, 1.0)
-  solfw3, ll3 = MSDE.forwardguiding(kg3, message, (x0, ll0), Z, inplace=false)
+  solfw3, ll3 = MSDE.forwardguiding(kg3, message, (x0, ll0), EM(false), Z, inplace=false)
   @test ll1 ≈ ll3
   @test isapprox(solfw1.u, solfw3.u, rtol=1e-14)
+end
+
+
+
+@testset "multivariate forward guidng tests" begin
+  seed = 12345
+  Random.seed!(seed)
+  d = 2
+  u0 = randn(2)
+  θlin = (randn(d,d), randn(d), Diagonal([0.1, 0.1]))
+
+  Σ(θ) = Diagonal(θ[2]) # just to generate the noise_rate_prototype
+
+  f(u,p,t) = p[1]*u + p[2]
+  gvec(u,p,t) = diag(p[3])
+  g(u,p,t) = p[3]
+  function gstepvec!(dx, _, u, p, t, dw, _)
+    dx .+= diag(p[3]).*dw
+  end
+
+  function gstep!(dx, _, u, p, t, dw, _)
+    dx .+= p[3]*dw
+  end
+
+  # Define a sparse matrix by making a dense matrix and setting some values as not zero
+  A = zeros(2,2)
+  A[1,1] = 1
+  A[2,2] = 1
+  A = sparse(A)
+
+  # time range
+  tstart = 0.0
+  tend = 1.0
+  dt = 0.02
+  trange = tstart:dt:tend
+
+  # define kernels
+  k1 = MSDE.SDEKernel(f,gvec,trange,θlin)
+  k2 = MSDE.SDEKernel(f,g,trange,θlin,Σ(θlin))
+  k3 = MSDE.SDEKernel(Mitosis.AffineMap(θlin[1], θlin[2]), Mitosis.ConstantMap(θlin[3]), trange, θlin, Σ(θlin))
+
+
+  sol, solend = MSDE.sample(k1, u0, EM(false), save_noise=true)
+  v = solend
+  c = randn()
+  Pmat = randn(d,d)
+
+  message1, backward1 = MSDE.backwardfilter(k1, WGaussian{(:μ, :Σ, :c)}(v, Pmat, c))
+  message2, backward2 = MSDE.backwardfilter(k2, WGaussian{(:μ, :Σ, :c)}(v, Pmat, c))
+  message3, backward3 = MSDE.backwardfilter(k3, WGaussian{(:μ, :Σ, :c)}(v, Pmat, c))
+
+  @test message1.soldis == message2.soldis
+  @test message2.soldis == message3.soldis
+
+  ll0 = randn()
+
+  @testset "StochasticDiffEq EM() solver" begin
+    sol1, ll1 = MSDE.forwardguiding(k1, message1, (u0, ll0), EM(false); save_noise=true)
+    Z = pCN(sol1.W, 1.0)
+    sol2, ll2 = MSDE.forwardguiding(k2, message2, (u0, ll0), EM(false), Z; save_noise=true)
+    Z = pCN(sol1.W, 1.0)
+    sol3, ll3 = MSDE.forwardguiding(k3, message3, (u0, ll0), EM(false), Z; save_noise=true)
+
+    @test sol1.u ≈ sol2.u rtol=1e-14
+    @test sol2.u ≈ sol3.u rtol=1e-14
+    @test ll1 == ll2
+    @test ll2 == ll3
+  end
+
+  @testset "internal solver" begin
+    @testset "without passing a noise" begin
+      Random.seed!(seed)
+      sol1, ll1 = MSDE.forwardguiding(k1, message1, (u0, ll0), MSDE.EulerMaruyama!())
+      Random.seed!(seed)
+      sol2, ll2 = MSDE.forwardguiding(k2, message2, (u0, ll0), MSDE.EulerMaruyama!())
+      Random.seed!(seed)
+      sol3, ll3 = MSDE.forwardguiding(k3, message3, (u0, ll0), MSDE.EulerMaruyama!())
+      @test ll1 == ll2
+      @test ll1 == ll3
+      @test sol1 == sol2
+      @test sol1 == sol3
+    end
```

Comment:
tests are not complete yet. Need to make forwardguiding run first. (reinterpret step from above fails at the moment)